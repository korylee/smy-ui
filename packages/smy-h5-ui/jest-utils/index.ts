import Vue from 'vue'
import { Wrapper } from '@vue/test-utils'
import { IN_BROWSER } from '@smy-h5/shared'

Vue.config.productionTip = false

export const delay = (time = 0) => new Promise((resolve) => setTimeout(resolve, time))

export function getTouch(el: Element | Window | Document, x: number, y: number) {
  return {
    identifier: Date.now(),
    target: el,
    pageX: x,
    pageY: y,
    clientX: x,
    clientY: y,
    radiusX: 2.5,
    radiusY: 2.5,
    rotationAngle: 10,
    force: 0.5,
  }
}

export function trigger(
  wrapper: Wrapper<Vue> | HTMLElement | Window | Document,
  eventName: string,
  x = 0,
  y = 0,
  offsetX = 0,
  offsetY = 0,
) {
  const el = 'element' in wrapper ? wrapper.element : wrapper
  const touchList = [getTouch(el, x, y)]

  const event = document.createEvent('CustomEvent')
  event.initCustomEvent(eventName, true, true, {})

  Object.assign(event, {
    clientX: x,
    clientY: y,
    offsetX,
    offsetY,
    touches: touchList,
    targetTouches: touchList,
    changedTouches: touchList,
  })
  el.dispatchEvent(event)

  return Vue.nextTick()
}

export async function triggerDrag(el: HTMLElement, relativeX: number, relativeY: number) {
  let x = relativeX
  let y = relativeY
  let startX = 0
  let startY = 0
  if (relativeX < 0) {
    startX = Math.abs(relativeX)
    x = 0
  }
  if (relativeY < 0) {
    startY = Math.abs(relativeY)
    y = 0
  }
  await trigger(el, 'touchstart', startX, startY)
  await trigger(el, 'touchmove', x / 4, y / 4)
  await trigger(el, 'touchmove', x / 3, y / 3)
  await trigger(el, 'touchmove', x / 2, y / 2)
  await trigger(el, 'touchmove', x, y)
  await delay(300)
  await trigger(el, 'touchend', x, y)
}

export function mockConsole(method: keyof Console, fn: any = () => {}) {
  const originMethod = console[method] as any
  console[method] = fn
  return () => {
    console[method] = originMethod
  }
}

export function mockGetBoundingClientRect(rect: Partial<DOMRect>): () => void {
  if (IN_BROWSER) {
    const spy = jest.spyOn(Element.prototype, 'getBoundingClientRect')
    spy.mockReturnValue(rect as DOMRect)
    return () => spy.mockRestore()
  }
  return () => {}
}
