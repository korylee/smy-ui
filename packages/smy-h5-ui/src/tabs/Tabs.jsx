import { getListener } from '../_mixins/listeners'
import { createParentMixin } from '../_mixins/relation'
import { useWindowSize } from '../_utils/composable/useWindowSize'
import {
  isNil,
  convertToUnit,
  doubleRaf,
  getElementTop,
  getParentScroller,
  getRect,
  setRootScrollTop,
  toPxNum,
} from '@smy-h5/shared'
import { createNamespace } from '../_utils/vue/create'
import SmySticky from '../sticky'
import SmySwiper from '../swiper'
import { scrollLeftTo, scrollTopTo } from './utils'
import Intersect from '../intersect'
import { props } from './props'
import { onMountedOrActivated } from '../_utils/vue/lifetime'
import { PopupMixin } from '../popup/provide'
import { TABS_KEY } from './shared'

import '../sticky/sticky.less'
import './tabs.less'

const [name, bem] = createNamespace('tabs')

const getTabName = (tab, index) => tab.name ?? index

export default {
  name,
  props,
  directives: { Intersect },
  mixins: [createParentMixin(TABS_KEY), PopupMixin],
  data: () => ({
    inited: false,
    currentIndex: -1,
    tabHeight: 0,
    lineStyle: {},
  }),
  computed: {
    scrollable({ ellipsis, shrink, children, scrollThreshold }) {
      return children.length > +scrollThreshold || !ellipsis || shrink
    },
    currentName({ children, currentIndex }) {
      const activeTab = children[currentIndex]
      if (!activeTab) return
      return getTabName(activeTab, currentIndex)
    },
    navStyle({ color, background }) {
      return { borderColor: color, background }
    },
    offsetTopPx({ offsetTop }) {
      return toPxNum(offsetTop)
    },
    scrollOffset({ sticky, tabHeight, offsetTopPx }) {
      if (sticky) {
        return tabHeight + offsetTopPx
      }
      return 0
    },
  },
  watch: {
    active(value) {
      if (value === this.currentName) return
      this.setCurrentIndexByName(value)
    },
    children() {
      if (!this.inited) return
      this.setCurrentIndexByName(this.active)
      this.setLine()
      this.$nextTick(() => {
        this.scrollIntoView(true)
      })
    },
  },
  created() {
    const vm = this
    onMountedOrActivated(vm, vm.init)
    const windowSize = useWindowSize()
    this.$watch(() => windowSize.width, vm.resize)

    const add = () => {
      window.addEventListener('pageshow', vm.resize)
      doubleRaf(() => {
        vm.scroller = getParentScroller(vm.$refs.root)
        vm.scroller.addEventListener('scroll', vm.onScroll, { passive: true })
      })
    }
    const remove = () => {
      vm.scroller && vm.scroller.removeEventListener('scroll', vm.onScroll)
      window.removeEventListener('pageshow', vm.resize)
    }
    onMountedOrActivated(vm, add)
    vm.$on(['hook:beforeDestory', 'hook:deactivated'], remove)
    vm.popupProvider?.$on('show', vm.setLine)
  },
  methods: {
    onScroll() {
      if (!this.scrollspy || this.lockScroll) {
        return
      }
      const index = this.getCurrentIndexOnScroll()
      this.setCurrentIndex(index)
    },
    scrollTo(name) {
      this.$nextTick(() => {
        this.setCurrentIndexByName(name)
        this.scrollToCurrentContent(true)
      })
    },
    onClickTab(item, index, event) {
      const { title, disabled } = item
      const onBeforeChange = getListener.call(this, 'before-change')
      const onClickTab = getListener.call(this, 'click-tab')
      const name = getTabName(item, index)
      if (!disabled) {
        Promise.resolve(onBeforeChange ? onBeforeChange(name) : true).then(() => {
          this.setCurrentIndex(index)
          this.scrollToCurrentContent()
        })
      }
      onClickTab?.({ name, title, event, disabled })
    },
    findAvailableTabIndex(index) {
      const diff = index < this.currentIndex ? -1 : 1
      const { children } = this
      while (index >= 0 && index < children.length) {
        if (!children[index].disabled) {
          return index
        }
        index += diff
      }
    },
    scrollIntoView(immediate) {
      const { nav } = this.$refs
      const { currentIndex, scrollable, children } = this
      const item = children[currentIndex]

      if (!scrollable || !nav || !children.length || !item) {
        return
      }
      const title = item.$refs['tab-title']?.$el
      if (!title) {
        return
      }
      const to = title.offsetLeft - (nav.offsetWidth - title.offsetWidth) / 2
      if (this.cancelScrollLeftToRaf) this.cancelScrollLeftToRaf()
      this.cancelScrollLeftToRaf = scrollLeftTo(nav, to, immediate ? 0 : +this.duration)
    },
    setCurrentIndex(currentIndex, skipScrollIntoView) {
      const newIndex = this.findAvailableTabIndex(currentIndex)
      if (isNil(newIndex)) {
        return
      }
      const newTab = this.children[newIndex]
      const newName = getTabName(newTab, newIndex)
      const shouldEmitChange = this.currentIndex !== null
      if (this.currentIndex !== newIndex) {
        const { swiper } = this.$refs
        if (swiper?.to && swiper.active !== newIndex) {
          swiper.to?.(newIndex)
        }
        this.currentIndex = newIndex
        if (!skipScrollIntoView) {
          this.scrollIntoView()
        }
        this.setLine()
      }
      if (newName !== this.active) {
        this.$emit('update:active', newName)
        if (shouldEmitChange) {
          this.$emit('change', newName, newTab.title)
        }
      }

      const { root, sticky } = this.$refs
      if (sticky?.fixed && !this.scrollspy) {
        setRootScrollTop(Math.ceil(getElementTop(root) - this.offsetTopPx))
      }
    },
    setCurrentIndexByName(name, skipScrollIntoView = false) {
      const { children } = this
      const matchedIndex = children.findIndex((tab, index) => getTabName(tab, index) === name)
      const index = matchedIndex === -1 ? 0 : matchedIndex
      this.setCurrentIndex(index, skipScrollIntoView)
    },
    scrollToCurrentContent(immediate = false) {
      if (!this.scrollspy) return
      const { children, currentIndex, scrollOffset, scroller } = this
      const item = children[currentIndex]
      const target = item?.$el
      if (!target || !scroller) return
      const to = getElementTop(target, scroller) - scrollOffset
      this.lockScroll = true
      if (this.cancelScrollTopToRaf) this.cancelScrollTopToRaf()
      this.cancelScrollTopToRaf = scrollTopTo(scroller, to, immediate ? 0 : +this.duration, () => {
        this.lockScroll = false
      })
    },
    setLine() {
      const shouldAnimate = this.inited
      this.$nextTick(() => {
        const { children, currentIndex, lineWidth, lineHeight, color, duration } = this
        const item = children[currentIndex]
        const title = item.$refs['tab-title']?.$el
        if (!title) {
          this.lineStyle = {
            display: 'none',
          }
          return
        }
        const left = title.offsetLeft + title.offsetWidth / 2
        let height
        let borderRadius
        if (!isNil(lineHeight)) {
          height = convertToUnit(lineHeight)
          borderRadius = height
        }
        this.lineStyle = {
          width: convertToUnit(lineWidth),
          backgroundColor: color,
          transform: `translateX(${left}px) translateX(-50%)`,
          transitionDuration: shouldAnimate ? `${duration}ms` : undefined,
          borderRadius,
          height,
        }
      })
    },
    getCurrentIndexOnScroll() {
      const { children, scrollOffset } = this
      for (let index = 0; index < children.length; index++) {
        const { top } = getRect(children[index]?.$el)
        if (top > scrollOffset) {
          return !index ? 0 : index - 1
        }
      }
      return children.length - 1
    },
    resize() {
      this.setLine()
      this.$nextTick(() => {
        this.scrollIntoView(true)
        this.$refs.content?.$refs.swiper?.resize()
      })
    },
    init() {
      this.setCurrentIndexByName(this.active, true)
      this.$nextTick(() => {
        this.inited = true
        const { wrap } = this.$refs
        this.tabHeight = getRect(wrap).height

        this.scrollIntoView(true)
      })
    },
  },
  render() {
    const vm = this
    const _c = vm.$createElement
    const renderTitle = () => {
      const { children: tabs, shrink, scrollable } = vm
      const children = [
        <div ref="wrap" class={bem('wrap')}>
          <div
            ref="nav"
            class={bem('nav', { line: true, shrink, scrollable })}
            style={vm.navStyle}
            role="tablist"
            aria-orientation="horizontal"
          >
            {vm._t('nav-left')}
            {vm._l(tabs, (tab, index) =>
              tab.renderTitle?.({
                attrs: {
                  scrollable,
                  shrink,
                },
                listeners: { click: ($event) => vm.onClickTab(tab, index, $event) },
              }),
            )}
            {tabs.length ? <div class={bem('line')} style={vm.lineStyle}></div> : null}
            {vm._t('nav-right')}
          </div>
        </div>,
        vm._t('nav-bottom'),
      ]
      return vm.sticky
        ? _c(SmySticky, { ref: 'sticky', on: { scroll: (event) => vm.$emit('scroll', event) } }, children)
        : children
    }
    const renderContent = () => {
      const { swipeable, animated: _animated } = vm
      const animated = _animated || swipeable
      const { duration } = vm
      const children = vm._t('default')
      return (
        <div class={bem('content', { animated })}>
          {animated
            ? _c(
                SmySwiper,
                {
                  ref: 'swiper',
                  attrs: { loop: false, indicators: false, duration, touchable: swipeable },
                  on: { change: (activeIndex) => vm.setCurrentIndex(activeIndex) },
                  class: bem('track'),
                },
                children,
              )
            : children}
        </div>
      )
    }

    return (
      <div ref="root" v-intersect={vm.setLine} class={bem()}>
        {renderTitle()}
        {renderContent()}
      </div>
    )
  },
}
