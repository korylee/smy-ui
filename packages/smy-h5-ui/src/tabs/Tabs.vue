<template>
  <div ref="root" v-intersect="setLine" :class="bem()">
    <maybe-sticky ref="sticky" :maybe="sticky" :container="$refs.root" @scroll="$emit('scroll', $event)">
      <div ref="wrap" :class="bem('wrap')">
        <div
          ref="nav"
          :class="bem('nav', { line: true, shrink, scrollable })"
          :style="navStyle"
          role="tablist"
          aria-orientation="horizontal"
        >
          <slot name="nav-left"></slot>
          <tab-title
            v-for="(item, index) of children"
            :key="item._uid"
            :tab="item"
            :shrink="shrink"
            :scrollable="scrollable"
            @click="onClickTab(item, index, $event)"
          />
          <div v-if="children.length" :class="bem('line')" :style="lineStyle"></div>
          <slot name="nav-right"></slot>
        </div>
      </div>
      <slot name="nav-bottom"></slot>
    </maybe-sticky>
    <tabs-content
      ref="content"
      :count="children.length"
      :animated="animated"
      :swipeable="swipeable"
      :duration="duration"
      :inited="inited"
      :current-index="currentIndex"
      @change="setCurrentIndex"
    >
      <slot />
    </tabs-content>
  </div>
</template>

<script>
import { ListenersMixin } from '../_mixins/listeners'
import { createParentMixin } from '../_mixins/relation'
import { useWindowSize } from '../_utils/composable/useWindowSize'
import {
  convertToUnit,
  doubleRaf,
  getElementTop,
  getParentScroller,
  getRect,
  setRootScrollTop,
  toPxNum,
} from '../_utils/dom'
import { isNil } from '../_utils/is'
import { createMaybeComponent } from '../_utils/vue/component'
import { createNamespace } from '../_utils/vue/create'
import SmySticky from '../sticky'
import { scrollLeftTo, scrollTopTo } from './utils'
import Intersect from '../intersect'
import TabsContent from './TabsContent.vue'
import { props } from './props'
import { onMountedOrActivated } from '../_utils/vue/lifetime'

const [name, bem] = createNamespace('tabs')

const MaybeSticky = createMaybeComponent(SmySticky)

const TabTitle = {
  functional: true,
  props: { tab: Object },
  render(h, context) {
    const { props } = context
    return props.tab?.renderTitle(context)
  },
}

const getTabName = (tab, index) => tab.name ?? index

export default {
  name,
  components: { MaybeSticky, TabTitle, TabsContent },
  props,
  directives: { Intersect },
  mixins: [createParentMixin('tabs'), ListenersMixin],
  data: () => ({
    inited: false,
    currentIndex: -1,
    tabHeight: 0,
    windowSize: useWindowSize(),
    lineStyle: {},
  }),
  computed: {
    scrollable({ ellipsis, shrink, children, scrollThreshold }) {
      return children.length > +scrollThreshold || !ellipsis || shrink
    },
    currentName({ children, currentIndex }) {
      const activeTab = children[currentIndex]
      if (!activeTab) return
      return getTabName(activeTab, currentIndex)
    },
    navStyle({ color, background }) {
      return { borderColor: color, background }
    },
    offsetTopPx({ offsetTop }) {
      return toPxNum(offsetTop)
    },
    scrollOffset({ sticky, tabHeight, offsetTopPx }) {
      if (sticky) {
        return tabHeight + offsetTopPx
      }
      return 0
    },
  },
  watch: {
    active(value) {
      if (value === this.currentName) return
      this.setCurrentIndexByName(value)
    },
    children() {
      if (!this.inited) return
      this.setCurrentIndexByName(this.active)
      this.setLine()
      this.$nextTick(() => {
        this.scrollIntoView(true)
      })
    },
    'windowSize.width': 'resize',
  },
  created() {
    onMountedOrActivated(this, this.init)

    const add = () => {
      doubleRaf(() => {
        this.scroller = getParentScroller(this.$refs.root)
        this.scroller.addEventListener('scroll', this.onScroll, { passive: true })
      })
    }
    const remove = () => {
      this.scroller && this.scroller.removeEventListener('scroll', this.onScroll)
    }
    onMountedOrActivated(this, add)
    this.$on(['hook:beforeDestory', 'hook:deactivated'], remove)
  },
  methods: {
    bem,
    onScroll() {
      if (!this.scrollspy || this.lockScroll) {
        return
      }
      const index = this.getCurrentIndexOnScroll()
      this.setCurrentIndex(index)
    },
    scrollTo(name) {
      this.$nextTick(() => {
        this.setCurrentIndexByName(name)
        this.scrollToCurrentContent(true)
      })
    },
    onClickTab(item, index, event) {
      const { title, disabled } = item
      const { getListener } = this
      const onBeforeChange = getListener('before-change')
      const onClickTab = getListener('click-tab')
      const name = getTabName(item, index)
      if (!disabled) {
        Promise.resolve(onBeforeChange ? onBeforeChange(name) : true).then(() => {
          this.setCurrentIndex(index)
          this.scrollToCurrentContent()
        })
      }
      onClickTab?.({ name, title, event, disabled })
    },
    findAvailableTabIndex(index) {
      const diff = index < this.currentIndex ? -1 : 1
      const { children } = this
      while (index >= 0 && index < children.length) {
        if (!children[index].disabled) {
          return index
        }
        index += diff
      }
    },
    scrollIntoView(immediate) {
      const { nav } = this.$refs
      const { currentIndex, scrollable, children } = this
      const item = children[currentIndex]

      if (!scrollable || !nav || !children.length || !item) {
        return
      }
      const title = item.$refs['tab-title']?.$el
      if (!title) {
        return
      }
      const to = title.offsetLeft - (nav.offsetWidth - title.offsetWidth) / 2
      if (this.cancelScrollLeftToRaf) this.cancelScrollLeftToRaf()
      this.cancelScrollLeftToRaf = scrollLeftTo(nav, to, immediate ? 0 : +this.duration)
    },
    setCurrentIndex(currentIndex, skipScrollIntoView) {
      const newIndex = this.findAvailableTabIndex(currentIndex)
      if (isNil(newIndex)) {
        return
      }
      const newTab = this.children[newIndex]
      const newName = getTabName(newTab, newIndex)
      const shouldEmitChange = this.currentIndex !== null
      if (this.currentIndex !== newIndex) {
        this.currentIndex = newIndex
        if (!skipScrollIntoView) {
          this.scrollIntoView()
        }
        this.setLine()
      }
      if (newName !== this.active) {
        this.$emit('update:active', newName)
        if (shouldEmitChange) {
          this.$emit('change', newName, newTab.title)
        }
      }

      const { root, sticky } = this.$refs
      if (sticky?.fixed && !this.scrollspy) {
        setRootScrollTop(Math.ceil(getElementTop(root) - this.offsetTopPx))
      }
    },
    setCurrentIndexByName(name, skipScrollIntoView = false) {
      const { children } = this
      const matchedIndex = children.findIndex((tab, index) => getTabName(tab, index) === name)
      const index = matchedIndex === -1 ? 0 : matchedIndex
      this.setCurrentIndex(index, skipScrollIntoView)
    },
    scrollToCurrentContent(immediate = false) {
      if (!this.scrollspy) return
      const { children, currentIndex, scrollOffset, scroller } = this
      const item = children[currentIndex]
      const target = item?.$el
      if (!target || !scroller) return
      const to = getElementTop(target, scroller) - scrollOffset
      this.lockScroll = true
      if (this.cancelScrollTopToRaf) this.cancelScrollTopToRaf()
      this.cancelScrollTopToRaf = scrollTopTo(scroller, to, immediate ? 0 : +this.duration, () => {
        this.lockScroll = false
      })
    },
    setLine() {
      const shouldAnimate = this.inited
      this.$nextTick(() => {
        const { children, currentIndex, lineWidth, lineHeight, color, duration } = this
        const item = children[currentIndex]
        const title = item?.$refs['tab-title']?.$el
        if (!title) {
          this.lineStyle = {
            display: 'none',
          }
          return
        }
        const left = title.offsetLeft + title.offsetWidth / 2
        let height
        let borderRadius
        if (!isNil(lineHeight)) {
          height = convertToUnit(lineHeight)
          borderRadius = height
        }
        this.lineStyle = {
          width: convertToUnit(lineWidth),
          backgroundColor: color,
          transform: `translateX(${left}px) translateX(-50%)`,
          transitionDuration: shouldAnimate ? `${duration}ms` : undefined,
          borderRadius,
          height,
        }
      })
    },
    getCurrentIndexOnScroll() {
      const { children, scrollOffset } = this
      for (let index = 0; index < children.length; index++) {
        const { top } = getRect(children[index]?.$el)
        if (top > scrollOffset) {
          return !index ? 0 : index - 1
        }
      }
      return children.length - 1
    },
    resize() {
      this.setLine()
      this.$nextTick(() => {
        this.scrollIntoView(true)
        this.$refs.content?.$refs.swipe?.resize()
      })
    },
    init() {
      this.setCurrentIndexByName(this.active, true)
      this.$nextTick(() => {
        this.inited = true
        const { wrap } = this.$refs
        this.tabHeight = getRect(wrap).height

        this.scrollIntoView(true)
      })
    },
  },
}
</script>

<style lang="less">
@import '../sticky/sticky.less';
@import './tabs.less';
</style>
