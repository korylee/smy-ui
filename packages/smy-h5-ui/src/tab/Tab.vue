<template>
  <smy-swiper-item
    v-if="tabs.animated || tabs.swipeable"
    :class="bem('panel-wrapper')"
    :tabindex="active ? 0 : -1"
    role="tabpanel"
  >
    <div :class="bem('panel')">
      <slot />
    </div>
  </smy-swiper-item>
  <div v-else v-show="tabs.scrollspy || active" :class="bem('panel')" role="tabpanel">
    <slot v-if="inited || tabs.scrollspy" />
  </div>
</template>
<script>
import { createChildrenMixin } from '../_mixins/relation'
import TabTitle from './TabTitle.vue'
import { assign, pick } from '@smy-h5/shared'
import { props } from './props'
import SmySwiperItem from '../swiper-item'
import { name, bem } from './utils'
import { TABS_KEY } from '../tabs/shared'

export default {
  name,
  props,
  mixins: [createChildrenMixin(TABS_KEY)],
  components: { SmySwiperItem },
  data: () => ({
    inited: false,
  }),
  computed: {
    active({ name, index, [TABS_KEY]: parent }) {
      const isActive = (name || index) === parent.currentName
      if (isActive && !this.inited) {
        this.init()
      }
      return isActive
    },
  },
  watch: {
    title() {
      const { [TABS_KEY]: parent } = this
      parent.setLine()
      parent.scrollIntoView()
    },
  },
  methods: {
    bem,
    renderTitle({ attrs: _attrs, listeners }) {
      const vm = this
      const { $createElement: h, $props, active, titleClass, titleStyle } = vm
      const attrs = assign(
        {
          role: 'tab',
          active,
        },
        _attrs,
        pick($props, ['badge', 'title', 'disabled']),
      )
      const mergedData = {
        on: listeners,
        attrs,
        class: titleClass,
        style: titleStyle,
        ref: 'tab-title',
      }
      return h(TabTitle, mergedData, [vm._t('title')])
    },
    init() {
      this.inited = true
      const { [TABS_KEY]: parent, name, index } = this
      if (parent.lazyRender) {
        parent.$emit('rendered', name || index)
      }
    },
  },
}
</script>
