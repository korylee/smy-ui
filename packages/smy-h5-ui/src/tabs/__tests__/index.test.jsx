import Tabs from '..'
import Tab from '../../tab'
import { mount } from '@vue/test-utils'
import { delay } from '../../../jest-utils'

test('should emit clickTab event when tab is clicked', async () => {
  const onClickTab = jest.fn()
  const wrapper = mount({
    render() {
      return (
        <Tabs onClickTab={onClickTab}>
          <Tab title="title1"></Tab>
          <Tab title="title2"></Tab>
        </Tabs>
      )
    },
  })
  await delay()
  const tabs = wrapper.findAll('.smy-tab')
  await tabs.at(0).trigger('click')
  expect(onClickTab).toHaveBeenCalledWith(
    expect.objectContaining({
      name: 0,
      title: 'title1',
      disabled: false,
    }),
  )
})

test('should switch tab after click the tab title', async () => {
  const onChange = jest.fn()
  const wrapper = mount({
    render() {
      return (
        <Tabs onChange={onChange}>
          <Tab title="title1">Text</Tab>
          <Tab title="title2">Text</Tab>
          <Tab title="title3" disabled>
            Text
          </Tab>
        </Tabs>
      )
    },
  })
  await delay()
  expect(wrapper.html()).toMatchSnapshot()

  const tabs = wrapper.findAll('.smy-tab')
  await tabs.at(1).trigger('click')
  await tabs.at(2).trigger('click')
  await delay()
  expect(wrapper.html()).toMatchSnapshot()
  expect(onChange).toHaveBeenCalledTimes(1)
})
