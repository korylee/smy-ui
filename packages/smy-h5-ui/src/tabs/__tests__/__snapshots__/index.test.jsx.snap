// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should switch tab after click the tab title 1`] = `
"<div class="smy-tabs">
  <div class="smy-tabs__wrap">
    <div role="tablist" aria-orientation="horizontal" class="smy-tabs__nav smy-tabs__nav--line">
      <div role="tab" tabindex="0" aria-selected="true" class="smy-tab smy-tab--active"><span class="smy-tab__text">title1</span></div>
      <div role="tab" tabindex="-1" class="smy-tab"><span class="smy-tab__text">title2</span></div>
      <div role="tab" aria-disabled="true" class="smy-tab smy-tab--disabled"><span class="smy-tab__text">title3</span></div>
      <div class="smy-tabs__line" style="transform: translateX(0px) translateX(-50%);"></div>
    </div>
  </div>
  <div class="smy-tabs__content">
    <div role="tabpanel" class="smy-tab__panel" style="">Text</div>
    <div role="tabpanel" class="smy-tab__panel" style="display: none;">
      <!---->
    </div>
    <div role="tabpanel" class="smy-tab__panel" style="display: none;">
      <!---->
    </div>
  </div>
</div>"
`;

exports[`should switch tab after click the tab title 2`] = `
"<div class="smy-tabs">
  <div class="smy-tabs__wrap">
    <div role="tablist" aria-orientation="horizontal" class="smy-tabs__nav smy-tabs__nav--line">
      <div role="tab" tabindex="-1" class="smy-tab"><span class="smy-tab__text">title1</span></div>
      <div role="tab" tabindex="0" class="smy-tab smy-tab--active" aria-selected="true"><span class="smy-tab__text">title2</span></div>
      <div role="tab" aria-disabled="true" class="smy-tab smy-tab--disabled"><span class="smy-tab__text">title3</span></div>
      <div class="smy-tabs__line" style="transform: translateX(0px) translateX(-50%); transition-duration: 300ms;"></div>
    </div>
  </div>
  <div class="smy-tabs__content">
    <div role="tabpanel" class="smy-tab__panel" style="display: none;">Text</div>
    <div role="tabpanel" class="smy-tab__panel" style="">Text</div>
    <div role="tabpanel" class="smy-tab__panel" style="display: none;">
      <!---->
    </div>
  </div>
</div>"
`;
